internal/dynamo/keys.go

```diff
@@ -2,107 +2,239 @@ package dynamo
 
 import (
 	"encoding/base64"
-	"fmt"
+	"errors"
+	"strings"
+)
 
-	"github.com/ministryofjustice/opg-modernising-lpa/internal/actor"
+const (
+	lpaPrefix                      = "LPA"
+	donorPrefix                    = "DONOR"
+	subPrefix                      = "SUB"
+	attorneyPrefix                 = "ATTORNEY"
+	certificateProviderPrefix      = "CERTIFICATE_PROVIDER"
+	documentPrefix                 = "DOCUMENT"
+	evidenceReceivedPrefix         = "EVIDENCE_RECEIVED"
+	organisationPrefix             = "ORGANISATION"
+	memberPrefix                   = "MEMBER"
+	memberInvitePrefix             = "MEMBERINVITE"
+	memberIDPrefix                 = "MEMBERID"
+	metadataPrefix                 = "METADATA"
+	donorSharePrefix               = "DONORSHARE"
+	donorInvitePrefix              = "DONORINVITE"
+	certificateProviderSharePrefix = "CERTIFICATEPROVIDERSHARE"
+	attorneySharePrefix            = "ATTORNEYSHARE"
 )
 
+func readKey(s string) (any, error) {
+	prefix, _, ok := strings.Cut(s, "#")
+	if !ok {
+		return nil, errors.New("malformed key")
+	}
+
+	switch prefix {
+	case lpaPrefix:
+		return LpaKeyType(s), nil
+	case donorSharePrefix:
+		return DonorShareKeyType(s), nil
+	case certificateProviderSharePrefix:
+		return CertificateProviderShareKeyType(s), nil
+	case attorneySharePrefix:
+		return AttorneyShareKeyType(s), nil
+	case donorPrefix:
+		return DonorKeyType(s), nil
+	case subPrefix:
+		return SubKeyType(s), nil
+	case attorneyPrefix:
+		return AttorneyKeyType(s), nil
+	case certificateProviderPrefix:
+		return CertificateProviderKeyType(s), nil
+	case documentPrefix:
+		return DocumentKeyType(s), nil
+	case evidenceReceivedPrefix:
+		return EvidenceReceivedKeyType(s), nil
+	case organisationPrefix:
+		return OrganisationKeyType(s), nil
+	case memberPrefix:
+		return MemberKeyType(s), nil
+	case memberInvitePrefix:
+		return MemberInviteKeyType(s), nil
+	case memberIDPrefix:
+		return MemberIDKeyType(s), nil
+	case metadataPrefix:
+		return MetadataKeyType(s), nil
+	case donorInvitePrefix:
+		return DonorInviteKeyType(s), nil
+	default:
+		return nil, errors.New("unknown key prefix")
+	}
+}
+
+type PK interface{ PK() string }
+
+type SK interface{ SK() string }
+
+type LpaKeyType string
+
+func (t LpaKeyType) PK() string { return string(t) }
+
 // LpaKey is used as the PK for all Lpa related information.
-func LpaKey(s string) string {
-	return "LPA#" + s
+func LpaKey(s string) LpaKeyType {
+	return LpaKeyType(lpaPrefix + "#" + s)
 }
 
+type DonorKeyType string
+
+func (t DonorKeyType) SK() string { return string(t) }
+func (t DonorKeyType) lpaOwner()  {}
```

Comment:
I get what this is doing but it may be confusing for future maintainers. I'm not sure if adding a comment here and on `OrganisationKeyType` or in [complex_keys](https://github.com/ministryofjustice/opg-modernising-lpa/pull/1198/files#diff-4a6072316e7cf0afa3680a57370a70f21fa5fbef4fd30d24bd5a83442bba0684R61) makes more sense but some pointer to its purpose may be helpful?